### ISSUE TITLE
Enable to "force" unlock exercises?

### ISSUE THREAD

**Issue by PragTob** [exercism/exercism#5978]
Hi there wonderful team,

first and foremost I want to thank you for your extremely good work making learning so much better for so many people. So, truly - thank you 💚 

I guess there is a strong opinion about this somewhere, but if there is I haven't found it via google or github - appologies.

I think it might be a good thing to be able to "force" unlock exercises. If there is, I haven't found it.

Why would I want this? Well, if you are already more advanced you don't necessarily want to spend the time doing the more "basic" exercises you already know how to do just to get where you want to be.

Example use cases:

* I know some $lang but I want to practice $topic (Elixir example: I want to learn about Linking for instance)
* I'm new to $lang but I already did the basics elsewhere before coming to exercism, I don't want to repeat the basics I just did (I know some repetition is good)
* I want to check out advanced exercises in $lang that I'm highly proficient in to see how much they'd teach someone about it whom I want to recommend exercism to (this is my case, I know I could check github but I'd like to also solve them)

Example of the amount of exercises one might need to do:

![Screenshot 2021-09-28 at 15 35 52](https://user-images.githubusercontent.com/606517/135100041-0e74ec8b-6dd6-4c2c-a1b4-887e4c59c6ae.png)

My suggestion would be, instead of stuff straight up not being unlockable/doable to have a popup that'd be like "Sure you want to unlock this exercise early? We strongly encourage you to follow the learning path" or to have some sort of setting that'd go "unlock all exercises".

I think this would be beneficial for learners and could hence widen the audience of exercism.

Again, thanks a lot for all your work. Go have a bunny:

![IMG_20180317_110254](https://user-images.githubusercontent.com/606517/135100510-660a1b45-2026-4ec1-bda6-2b9f748da9aa.jpg)

**Comment by iHiD** [comment]
Yo! Thanks for your support! 💙 

You can unlock all the practice exercises by switching to Practice Mode ("Disable Learning Mode" button on the `...` on the track page - nicely hidden away). However, that doesn't let you unlock Learning Exercises. Quoting myself from the issue Angelika linked to "The problem with letting people complete learning exercises in practice mode is that when they then switch back to learning mode (which people tend to do) then their concept trees are in really broken states"

I always find this a tricky area to get right. I was in v2 as well. I've spent ages considering it all over the years and not really come to any position that i like. Maybe when someone joins the track we give them the option of Learning or Practice mode to make this distinction clearer? I think we had this in v2 at the end (I've lost track though - feels like a lifetime ago! 🙂)

Open to thoughts! :)

Thanks for the bunny! 🐰

**Comment by marcandre** [comment]
Just wanted to chime in. Locking the exercises is a really *really* bad idea.
Learning Elixir, I did a single "easy" execise, then 2 "medium" ones, then a bunch of "hard" ones. I would not have had the patience to wait.

**Comment by iHiD** [comment]
Hi @marcandre. Firstly thanks for all your work in the Ruby ecosystem! Exercism wouldn't be here if it wasn't for people like you building the tech that underpins on.

> Locking the exercises is a really really bad idea.

I appreciate your opinion but I disagree. We've have had nearly 1m people use Exercism, and had hundreds of thousands of mentor discussions,  and we've learnt a huge amount from that, including that if you don't provide structure, people tend to not grasp the fundamentals of a language properly and instead start hacking around reusing their existing patterns and getting themselves into a real mess. Rather than learning the actual idioms that make Elixir (in this example) different from Ruby, they would just reuse their Rubyisms while writing Elixir and develop a whole set of behaviours that then need to be undone for them to really get good at that language. 

Fundamentally, if you're a decent developer it's easy to work through exercises in any language and complete them without trouble. But that's not what Exercism's about. Exercism is about getting you to think like an Elixir dev, not just write Elixir code. And that's why structure, and mentoring along with that works.

Of course there will be exceptions, and maybe (probably?) you're one of those people who has enough experience across languages to be able to do learn new idioms and put your Ruby-hat aside while doing Elxiir, and not need that structured pathway. For that 1% of people Practice Mode is there and available and will work great. But for the vast majority of people structure and learning pathways are important to actually get the idioms, which is why we have them.

> At minimum, clicking on a lock should hint at the "disable learning mode"

I agree we should probably make it easier to find Practice Mode. One problem with making it too available is that I suspect lots of people would reach for that and not actually take the path that's best for them. So while I agree with making it easier to find, I still want to strongly nudge people towards Learning Mode.

---

As I said in my comments to @PragTob, I think this is a complex issue to get right, and there's lots we can do to improve things. This issue/thread isn't by any means dead, it's just one of numerous improvements/challenges that all fight for my brainspace, and which are harder when they're product-level decisions because they have far-reaching consequences. Thanks for chiming in 🙂

**Comment by LandonSchropp** [comment]
Just to add another use case for supporting unlocking, my team is looking at using Exercism for group learning at work. We're a group of developers who are already proficient in the language, but we're looking to hone our skills, learn from each other and have fun doing programming exercises.

Without the ability to unlock exercises, Exercism is tough to use because we simply have too many people to ask everyone to tackle a large number of easy exercises.

**Comment by iHiD** [comment]
@LandonSchropp For your use case you can probably just switch to Practice Mode to unlock everything straight away.

**Comment by LandonSchropp** [comment]
@iHiD Ah, I didn't realize there was a practice mode. Thanks!

**Comment by marcandre** [comment]
Is the "Practice mode" still available? I'm on `/tracks/kotlin` and can't find it...

<img width="561" alt="image" src="https://github.com/exercism/exercism/assets/33770/87fa815c-da2f-47de-b013-0cc78ab01003">

**Comment by marcandre** [comment]
So I finished the Hello World, and now all the exercises are unlocked 🎉. Is that true for all tracks? If so, this issue can definitely be closed...

**Comment by angelikatyborska** [comment]
@marcandre No, that is not true for all tracks. Kotlin has only practice mode, where after doing "hello world" all practice exercises are unlocked. Some tracks have both learning mode and practice mode. In learning mode, you have to do learning exercises in order, and doing learning exercises unlocks some practice exercises. In every track with a learning mode, you can switch to practice mode which removes all the learning exercises and unlocks all practice exercises.

The original issue author's problem is about being able to unlock all learning exercises at once.
